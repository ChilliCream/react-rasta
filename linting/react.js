"use strict";

const ON = true;
const OFF = false;

module.exports = {
  extends: "tslint-react",
  rules: {
    // All
    "jsx-alignment": ON,
    "jsx-boolean-value": ON,
    "jsx-curly-spacing": [ON, "never"],
    "jsx-equals-spacing": [ON, "never"],
    "jsx-key": ON,
    "jsx-no-bind": ON,
    "jsx-no-lambda": OFF,
    "jsx-no-multiline-js": OFF,
    "jsx-no-string-ref": ON,
    "jsx-self-close": ON,
    "jsx-wrap-multiline": ON,
  },
};
