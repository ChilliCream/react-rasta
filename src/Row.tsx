// These imports are needed for declaration files
import * as _React from "react";
import * as _StyledComponents from "styled-components";
// -------------------------------------------------------------------
import { _getBreakpoints } from "./BreakpointMap";
import {
  BreakpointValue,
  BreakpointValues,
  PropertyValue
} from "./BreakpointValue";
import Column from "./Column";
import styled from "./StyledComponents";
import ThemeProperties, { Theme } from "./Theme";
import {
  _getGutterWidth,
  _ensureInjectingGlobal,
  _render,
  _css
} from "./Utilities";

export interface RowProperties extends ThemeProperties {
  noGutter?: BreakpointValue<boolean>;
}

const Row = styled.div`
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;

  ${(props: RowProperties) => {
    const breakpoints = _getBreakpoints(props!.theme);
    const renderer = {
      noGutter: (value?: PropertyValue) => _renderNoGutter(value as boolean)
    };
    const valueMap = {
      noGutter: props!.noGutter as BreakpointValues<boolean>
    };
    const width = _getGutterWidth(props.theme);

    return `
        margin-right: -${width}px;
        margin-left: -${width}px;

        ${_render(valueMap, breakpoints, renderer)}
        ${_renderNoGutter(props!.noGutter as boolean)}
      `;
  }};
`;

export default Row;

function _renderNoGutter(noGutter?: boolean): string {
  if (noGutter != null && typeof noGutter === "boolean" && noGutter) {
    return _css`
      margin-right: 0;
      margin-left: 0;
    
      > ${Column} {
        padding-right: 0;
        padding-left: 0;
      }
    `;
  }

  return "";
}

_ensureInjectingGlobal();
