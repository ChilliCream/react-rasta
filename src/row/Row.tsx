import "../utils/bootstrap";
import { ClassAttributes, HTMLAttributes } from "react";
import { StyledComponentClass } from "styled-components";
import renderAlignItems from "./renderAlignItems";
import renderDirection from "./renderDirection";
import renderGutter from "./renderGutter";
import renderJustifyContent from "./renderJustifyContent";
import renderWrap from "./renderWrap";
import RowAlignItems from "./RowAlignItems";
import RowDirection from "./RowDirection";
import RowJustifyContent from "./RowJustifyContent";
import RowProperties from "./RowProperties";
import RowWrap from "./RowWrap";
import { BreakpointValue, PropertyValue } from "../media";
import { styled, Theme } from "../theme";
import { render } from "../utils";

const Row = styled.div`
  display: flex;

  ${(props: RowProperties) => {
    const renderer = {
      alignItems: (value?: PropertyValue) =>
        renderAlignItems(value as RowAlignItems),
      justifyContent: (value?: PropertyValue) =>
        renderJustifyContent(value as RowJustifyContent),
      direction: (value?: PropertyValue) =>
        renderDirection(value as RowDirection),
      gutter: (value?: PropertyValue) => renderGutter(value as boolean),
      wrap: (value?: PropertyValue) => renderWrap(value as RowWrap)
    };
    const valueMap = {
      alignItems: props!.alignItems as BreakpointValue<RowAlignItems>,
      justifyContent: props!.justifyContent as BreakpointValue<
        RowJustifyContent
      >,
      direction: props!.direction as BreakpointValue<RowDirection>,
      gutter: props!.noGutter as BreakpointValue<boolean>,
      wrap: props!.wrap as BreakpointValue<RowWrap>
    };

    return render(valueMap, renderer, props!.theme);
  }};
`;

export default Row;
