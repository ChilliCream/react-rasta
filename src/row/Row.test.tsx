// Module mocks must mocked before importing modules
jest.mock("../utils/bootstrap", () => jest.fn());

import "jest-styled-components";
import { configure, shallow } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import styled from "styled-components";
import Row from "./Row";

configure({ adapter: new Adapter() });

describe("<Row />", () => {
  it("should match the snapshot (no properties set)", () => {
    // act
    const result = shallow(<Row />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (direction: 'row') ", () => {
    // act
    const result = shallow(<Row direction={"row"} />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (direction: { xs: 'row', sm: 'row-reverse', md: 'column' })", () => {
    // act
    const result = shallow(
      <Row direction={{ xs: "row", sm: "row-reverse", md: "column" }} />
    );

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (noGutter: true) ", () => {
    // act
    const result = shallow(<Row noGutter />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (noGutter: { xs: true, sm: false, md: true })", () => {
    // act
    const result = shallow(
      <Row noGutter={{ xs: true, sm: false, md: true }} />
    );

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (direction: 'column', noGutter: true)", () => {
    // arrange
    const offset = 3;
    const order = 3;
    const size = 9;

    // act
    const result = shallow(<Row direction={"column"} noGutter={true} />);

    // assert
    expect(result).toMatchSnapshot();
  });
});
