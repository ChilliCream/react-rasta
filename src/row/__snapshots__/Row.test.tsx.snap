// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<Row /> should match the snapshot (alignContent: 'flex-end', alignItems: 'baseline', direction: 'column', justifyContent: center, noGutter: true, wrap: 'nowrap') 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: flex-end !important;
  -ms-flex-line-pack: end !important;
  align-content: flex-end !important;
  -webkit-align-items: baseline !important;
  -webkit-box-align: baseline !important;
  -ms-flex-align: baseline !important;
  align-items: baseline !important;
  -webkit-box-pack: center !important;
  -webkit-justify-content: center !important;
  -ms-flex-pack: center !important;
  justify-content: center !important;
  -webkit-flex-direction: column !important;
  -ms-flex-direction: column !important;
  flex-direction: column !important;
  margin-right: 0;
  margin-left: 0;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 0;
  padding-left: 0;
}

<div
  className="c0"
  direction="column"
  wrap="nowrap"
/>
`;

exports[
  `<Row /> should match the snapshot (alignContent: 'space-between')  1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: space-between !important;
  -ms-flex-line-pack: space-between !important;
  align-content: space-between !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
/>
`;

exports[
  `<Row /> should match the snapshot (alignContent: { xs: 'space-around', sm: 'flex-end', md: 'flex-start' }) 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: space-around !important;
  -ms-flex-line-pack: space-around !important;
  align-content: space-around !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

@media (min-width:576px) {
  .c0 {
    -webkit-align-content: flex-end !important;
    -ms-flex-line-pack: end !important;
    align-content: flex-end !important;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-align-content: flex-start !important;
    -ms-flex-line-pack: start !important;
    align-content: flex-start !important;
  }
}

<div
  className="c0"
/>
`;

exports[`<Row /> should match the snapshot (alignItems: 'strech')  1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch !important;
  -webkit-box-align: stretch !important;
  -ms-flex-align: stretch !important;
  align-items: stretch !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
/>
`;

exports[
  `<Row /> should match the snapshot (alignItems: { xs: 'baseline', sm: 'center', md: 'flex-end' }) 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline !important;
  -webkit-box-align: baseline !important;
  -ms-flex-align: baseline !important;
  align-items: baseline !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

@media (min-width:576px) {
  .c0 {
    -webkit-align-items: center !important;
    -webkit-box-align: center !important;
    -ms-flex-align: center !important;
    align-items: center !important;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-align-items: flex-end !important;
    -webkit-box-align: flex-end !important;
    -ms-flex-align: flex-end !important;
    align-items: flex-end !important;
  }
}

<div
  className="c0"
/>
`;

exports[`<Row /> should match the snapshot (direction: 'row')  1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row !important;
  -ms-flex-direction: row !important;
  flex-direction: row !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
  direction="row"
/>
`;

exports[
  `<Row /> should match the snapshot (direction: { xs: 'row', sm: 'row-reverse', md: 'column' }) 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row !important;
  -ms-flex-direction: row !important;
  flex-direction: row !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

@media (min-width:576px) {
  .c0 {
    -webkit-flex-direction: row-reverse !important;
    -ms-flex-direction: row-reverse !important;
    flex-direction: row-reverse !important;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex-direction: column !important;
    -ms-flex-direction: column !important;
    flex-direction: column !important;
  }
}

<div
  className="c0"
  direction={
    Object {
      "md": "column",
      "sm": "row-reverse",
      "xs": "row",
    }
  }
/>
`;

exports[
  `<Row /> should match the snapshot (justifyContent: 'space-between')  1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify !important;
  -webkit-justify-content: space-between !important;
  -ms-flex-pack: justify !important;
  justify-content: space-between !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
/>
`;

exports[
  `<Row /> should match the snapshot (justifyContent: { xs: 'space-around', sm: 'flex-end', md: 'flex-start' }) 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around !important;
  -webkit-justify-content: space-around !important;
  -ms-flex-pack: space-around !important;
  justify-content: space-around !important;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

@media (min-width:576px) {
  .c0 {
    -webkit-box-pack: end !important;
    -webkit-justify-content: flex-end !important;
    -ms-flex-pack: end !important;
    justify-content: flex-end !important;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-box-pack: start !important;
    -webkit-justify-content: flex-start !important;
    -ms-flex-pack: start !important;
    justify-content: flex-start !important;
  }
}

<div
  className="c0"
/>
`;

exports[`<Row /> should match the snapshot (no properties set) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
/>
`;

exports[
  `<Row /> should match the snapshot (noGutter: { xs: true, sm: false, md: true }) 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0;
  margin-left: 0;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 0;
  padding-left: 0;
}

@media (min-width:576px) {
  .c0 {
    margin-right: -15px;
    margin-left: -15px;
  }

  .c0 > .sc-bdVaJa {
    padding-right: 15px;
    padding-left: 15px;
  }
}

@media (min-width:768px) {
  .c0 {
    margin-right: 0;
    margin-left: 0;
  }

  .c0 > .sc-bdVaJa {
    padding-right: 0;
    padding-left: 0;
  }
}

<div
  className="c0"
/>
`;

exports[`<Row /> should match the snapshot (noGutter: true)  1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0;
  margin-left: 0;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 0;
  padding-left: 0;
}

<div
  className="c0"
/>
`;

exports[`<Row /> should match the snapshot (wrap: 'wrap-reverse')  1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: wrap-reverse;
  -ms-flex-wrap: wrap-reverse;
  flex-wrap: wrap-reverse;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
  wrap="wrap-reverse"
/>
`;

exports[
  `<Row /> should match the snapshot (wrap: { xs: 'nowrap', sm: 'wrap-reverse', md: 'wrap' }) 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: -15px;
  margin-left: -15px;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c0 > .sc-bdVaJa {
  padding-right: 15px;
  padding-left: 15px;
}

@media (min-width:576px) {
  .c0 {
    -webkit-flex-wrap: wrap-reverse;
    -ms-flex-wrap: wrap-reverse;
    flex-wrap: wrap-reverse;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
  }
}

<div
  className="c0"
  wrap={
    Object {
      "md": "wrap",
      "sm": "wrap-reverse",
      "xs": "nowrap",
    }
  }
/>
`;
